open-apparel-registry/open-apparel-registry
# Support Front-end Environment Variables, Rollbar

## Overview

Adds a mechanism to allow for Django to inject environment variables into the front-end at run-time. These will override any set by react-scripts at compile-time. Usage will fallback to compile-time variables if a run-time version is not found. In the front-end, instead of using `process.env.REACT_APP_VARIABLE_NAME`, we should now use `env('REACT_APP_VARIABLE_NAME')`.

Also removes `django-spa`, which was a thin wrapper around an old version of `whitenoise`, with a more recent version of `whitenoise`. This is configured to serve static files found under `/static/` from the root URL `/`, mimicking the behavior of `django-spa`. Accessing the root directly `/` now goes through `oar.urls` and is handled by a new `web` app's `views.index`, which renders the react-scripts generated file with run-time variables injected. Accessing a URL not recognized by Django is assumed to be a front-end route, just as `django-spa` does, and is passed through to the front-end.

By convention, any environment variable available to Django that starts with `REACT_APP_` will be injected in to the front-end. Adding new environment variables is possible, although they have to be wired through many steps in Terraform, as done in the final commit.

Because these variables are populated via Django, to test them locally one must run `cibuild` and then go to [:8081/](http://localhost:8081/). When viewing the output of react-scripts at [:6543/](http://localhost:6543/) directly, these variables will not be available. However, their absence should not interfere with the development experience.

See commit messages for details.

Connects #64 
Connects #203 

## Demo

![image](https://user-images.githubusercontent.com/1430060/52834770-72825700-30b1-11e9-96f5-971b6f9fa7c6.png)

![image](https://user-images.githubusercontent.com/1430060/52834152-529d6400-30ae-11e9-9a72-04436589d7d9.png)

![image](https://user-images.githubusercontent.com/1430060/52834261-ec651100-30ae-11e9-8781-a5b1fcf5375e.png)

## Notes

Notably, the variables are only injected when the page is rendered via Django. In the current setup, WhiteNoise receives requests first, and if it can find a corresponding static file to return it will do so immediately. It is only those requests that do not match a static file that are sent on to Django. This means that while any request to `/` _will_ have the environment variables injected, any request to `/index.html` _will not_. I have not been able to think of a way around this short of removing WhiteNoise entirely, but perhaps this is an acceptable compromise at this stage.

## Testing Instructions

* Check out this branch
* Edit your `.env.backend` file to include the key `REACT_APP_ROLLBAR_POST_CLIENT_ITEM_ACCESS_TOKEN` with the `post_client_item` access token from Rollbar
* Run `scripts/update`, then `scripts/cibuild`, then `scripts/server`
* Go to [:8081/](http://localhost:8081/) and open the developer tools
* Trigger an error from the console

    ```
    window.onerror("TestError: Hello world", window.location.href)
    ```

* Log in to Rollbar and ensure you see the error there
* Go to [:6543/](http://localhost:6543/) and ensure development and hot-reloading work just as before, despite an extra error in the console for the invalid HTML of the template.
